File: packages/i18nify-js/src/modules/mandate/getMandateFrequencyLabel.ts
import { withErrorBoundary } from '../../common/errorBoundary';
import { MandateFrequencyCode, MandateFrequencyLabel } from './types';

// Display labels for recurring mandate frequency codes.
// `days` is the approximate number of days between recurrences (undefined for AS_PRESENTED).
// Lookup is case-insensitive. BI_WEEKLY is an alias for FORTNIGHTLY;
// SEMI_ANNUAL for HALF_YEARLY; ANNUAL for YEARLY.
const FREQUENCY_LABEL_MAP: Readonly<Record<string, MandateFrequencyLabel>> = {
  DAILY: {
    label: 'Daily',
    description: 'Recurring every day',
    days: 1,
  },
  WEEKLY: {
    label: 'Weekly',
    description: 'Recurring every week',
    days: 7,
  },
  FORTNIGHTLY: {
    label: 'Fortnightly',
    description: 'Recurring every two weeks',
    days: 14,
  },
  BI_WEEKLY: {
    label: 'Bi-Weekly',
    description: 'Recurring every two weeks',
    days: 14,
  },
  MONTHLY: {
    label: 'Monthly',
    description: 'Recurring every month',
    days: 30,
  },
  BI_MONTHLY: {
    label: 'Bi-Monthly',
    description: 'Recurring every two months',
    days: 60,
  },
  QUARTERLY: {
    label: 'Quarterly',
    description: 'Recurring every three months',
    days: 90,
  },
  HALF_YEARLY: {
    label: 'Half-Yearly',
    description: 'Recurring every six months',
    days: 180,
  },
  SEMI_ANNUAL: {
    label: 'Semi-Annual',
    description: 'Recurring every six months',
    days: 180,
  },
  YEARLY: {
    label: 'Yearly',
    description: 'Recurring every year',
    days: 365,
  },
  ANNUAL: {
    label: 'Annual',
    description: 'Recurring every year',
    days: 365,
  },
  AS_PRESENTED: {
    label: 'As Presented',
    description: 'Collected as and when presented by the merchant',
  },
};

const getMandateFrequencyLabel = (
  frequencyCode: MandateFrequencyCode | string,
): MandateFrequencyLabel => {
  if (!frequencyCode)
    throw new Error(
      `Parameter 'frequencyCode' is invalid! The received value was: ${frequencyCode}.`,
    );

  const definition =
    FREQUENCY_LABEL_MAP[(frequencyCode as string).toUpperCase()];
  if (!definition)
    throw new Error(
      `Frequency code "${frequencyCode}" is not supported. ` +
        `Supported codes: ${Object.keys(FREQUENCY_LABEL_MAP).join(', ')}.`,
    );

  return definition;
};

export default withErrorBoundary<typeof getMandateFrequencyLabel>(
  getMandateFrequencyLabel,
);

File: packages/i18nify-go/modules/banking/get_mandate_frequency_label.go
package banking

import (
	"fmt"
	"strings"
)

// MandateFrequencyLabel describes a recurring mandate frequency.
// Days is the approximate number of days between recurrences; 0 means not applicable (AS_PRESENTED).
type MandateFrequencyLabel struct {
	Label       string `json:"label"`
	Description string `json:"description"`
	Days        int    `json:"days,omitempty"`
}

// frequencyLabelMap maps uppercase frequency code keys to their display labels.
// BI_WEEKLY is an alias for FORTNIGHTLY; SEMI_ANNUAL for HALF_YEARLY; ANNUAL for YEARLY.
var frequencyLabelMap = map[string]MandateFrequencyLabel{
	"DAILY":       {Label: "Daily", Description: "Recurring every day", Days: 1},
	"WEEKLY":      {Label: "Weekly", Description: "Recurring every week", Days: 7},
	"FORTNIGHTLY": {Label: "Fortnightly", Description: "Recurring every two weeks", Days: 14},
	"BI_WEEKLY":   {Label: "Bi-Weekly", Description: "Recurring every two weeks", Days: 14},
	"MONTHLY":     {Label: "Monthly", Description: "Recurring every month", Days: 30},
	"BI_MONTHLY":  {Label: "Bi-Monthly", Description: "Recurring every two months", Days: 60},
	"QUARTERLY":   {Label: "Quarterly", Description: "Recurring every three months", Days: 90},
	"HALF_YEARLY": {Label: "Half-Yearly", Description: "Recurring every six months", Days: 180},
	"SEMI_ANNUAL": {Label: "Semi-Annual", Description: "Recurring every six months", Days: 180},
	"YEARLY":      {Label: "Yearly", Description: "Recurring every year", Days: 365},
	"ANNUAL":      {Label: "Annual", Description: "Recurring every year", Days: 365},
	"AS_PRESENTED": {
		Label:       "As Presented",
		Description: "Collected as and when presented by the merchant",
	},
}

// GetMandateFrequencyLabel returns display label information for the given
// mandate frequency code (e.g. "MONTHLY", "WEEKLY"). Lookup is case-insensitive.
func GetMandateFrequencyLabel(frequencyCode string) (MandateFrequencyLabel, error) {
	if strings.TrimSpace(frequencyCode) == "" {
		return MandateFrequencyLabel{}, fmt.Errorf("banking: GetMandateFrequencyLabel: frequencyCode must be a non-empty string")
	}
	label, ok := frequencyLabelMap[strings.ToUpper(strings.TrimSpace(frequencyCode))]
	if !ok {
		return MandateFrequencyLabel{}, fmt.Errorf("banking: GetMandateFrequencyLabel: frequency code %q is not supported", frequencyCode)
	}
	return label, nil
}
